### PR TITLE
Issue 135 : Added ability to limit number of builds to keep (just like Jenkins)

### PR DESCRIPTION
**Pull request by suratdas**
Addresses https://github.com/Visual-Regression-Tracker/Visual-Regression-Tracker/issues/135 
This pull request does following
1. Does not create separate images for baseline. It creates the reference to the actual image for baseline.
2. When deleting test run or test variation, it is first checked if it is in use by the other. Physical file is deleted only if it is not used by both variation and test run.
3. Since separate baseline files are not created, it saves disk space.

### CODE REVIEW

**Review by pashidlos — CHANGES_REQUESTED** [review]
@suratdas thanks for contribution! 

unfortunately this is not aligned with the idea of self contained entities 
having one object implicitly related to multiple objects makes complicated it's management and could potentially lead to broken relations
I would sacrifice this additional memory consumption to simplicity in logic

**Comment by suratdas** [comment]
I will modify to limit how many builds we want to keep just like Jenkins does without having us manually delete old builds. This is similar to issue https://github.com/Visual-Regression-Tracker/Visual-Regression-Tracker/issues/146
I plan to add code to this pull request.

File: src/builds/builds.controller.ts [hunk 1]
```diff
@@ -70,6 +70,7 @@ export class BuildsController {
   @UseGuards(ApiGuard)
   async create(@Body() createBuildDto: CreateBuildDto): Promise<BuildDto> {
     const project = await this.projectService.findOne(createBuildDto.project);
+    await this.buildsService.deleteOldBuilds(project.id, project.maxBuildAllowed);
```
Comment (pashidlos) [inline src/builds/builds.controller.ts:73]: not sure if this is the best place to make it
what about scheduled task?
https://docs.nestjs.com/techniques/task-scheduling

File: src/builds/builds.controller.spec.ts [hunk 2]
```diff
@@ -72,11 +74,13 @@ describe('Builds Controller', () => {
     const projectFindOneMock = jest.fn().mockResolvedValueOnce(project);
     const buildFindOrCreateMock = jest.fn().mockResolvedValueOnce(newBuild);
     const buildIncrementBuildNumberMock = jest.fn().mockResolvedValueOnce(buildWithNumber);
+    const deleteOldBuilds = jest.fn();
     controller = await initController({
       projectFindOneMock,
       buildFindOrCreateMock,
       buildIncrementBuildNumberMock,
       eventBuildCreatedMock,
+      deleteOldBuilds,
```
Comment (pashidlos) [inline src/builds/builds.controller.spec.ts:83]: add assertion method invoked

File: src/builds/builds.service.ts [hunk 3]
```diff
@@ -29,6 +29,19 @@ export class BuildsService {
     });
   }
 
+  async deleteOldBuilds(projectId: string, keepBuilds: number) {
+    const [total, buildList] = await Promise.all([
```
Comment (pashidlos) [inline src/builds/builds.service.ts:33]: by passing `undefined` in `findMany` method you could take the rest builds after skipping `keepBuilds`

File: src/builds/builds.service.ts [hunk 4]
```diff
@@ -29,6 +29,19 @@ export class BuildsService {
...
+      this.prismaService.build.count({ where: { projectId } }),
+      this.prismaService.build.findMany({
+        where: { projectId },
+        orderBy: { createdAt: 'desc' },
+      }),
+    ]);
+    for (let index = (keepBuilds - 1); index < total; index++) {
```
Comment (pashidlos) [inline src/builds/builds.service.ts:40]: could be replaced with for `forEach`?

File: src/builds/builds.service.ts [hunk 5]
```diff
@@ -29,6 +29,19 @@ export class BuildsService {
...
+      this.eventsGateway.buildDeleted(buildList[index]);
```
Comment (pashidlos) [inline src/builds/builds.service.ts:41]: didn't find where are the builds actually removed?

File: src/shared/events/events.gateway.ts [hunk 6]
```diff
@@ -48,6 +48,10 @@ export class EventsGateway {
     this.buildUpdated(testRun.buildId);
   }
 
+  buildDeleted(build: Build) {
```
Comment (pashidlos) [inline src/shared/events/events.gateway.ts:51]: better to send `BuildDto` to frontend

**Review by pashidlos — CHANGES_REQUESTED** [review]
could you add unit and/or e2e test for this feature?

File: src/builds/builds.controller.ts (hunk 1 above)
Comment (suratdas) [reply src/builds/builds.controller.ts:73]: Assuming most people are used to how Jenkins job deletes old builds on a build trigger, I had this behavior in mind and implemented it like that. If we are to implement some other mechanism, we can always have a relook with a future pull request. But, I think I would be happy with Jenkins-like behavior. Let me know what you think.

Comment (pashidlos) [reply src/builds/builds.controller.ts:73]: ok, make sense

File: src/builds/builds.service.ts (hunk 5 above)
Comment (suratdas) [reply src/builds/builds.service.ts:41]: Sorry, actually I was trying out with fronted code and forgot to include one line, fixed now.

File: src/builds/builds.service.ts [hunk 7]
```diff
@@ -29,6 +29,25 @@ export class BuildsService {
     });
   }
 
+  async deleteOldBuilds(projectId: string, keepBuilds: number) {
+    const [buildList] = await Promise.all([
```
Comment (pashidlos) [inline src/builds/builds.service.ts:33]: is that possible to reuse `this.findMany` method here?
you could just change it like this
`async findMany(projectId: string, take?: number, skip?: number): Promise<PaginatedBuildDto> {`

File: src/builds/builds.service.ts [hunk 8]
```diff
@@ -29,6 +29,25 @@ export class BuildsService {
...
+      this.prismaService.build.findMany({
+        where: { projectId },
+        orderBy: { createdAt: 'desc' },
+        take: undefined,
+        //Keep one build less because the new build will be added to this count to keep the max allowed build correct.
+        skip: (keepBuilds - 1)
+      }),
+    ]);
+    buildList.forEach(eachBuild => {
+      console.log("Deleting build" + JSON.stringify(eachBuild));
```
Comment (pashidlos) [inline src/builds/builds.service.ts:43]: change this to `this.logger`

File: src/builds/builds.service.ts [hunk 9]
```diff
@@ -29,6 +29,25 @@ export class BuildsService {
...
+      this.remove(eachBuild.id);
+      this.eventsGateway.buildDeleted(new BuildDto({
```
Comment (pashidlos) [inline src/builds/builds.service.ts:45]: could you make this as part of `this.remove` method?

**Review by pashidlos — CHANGES_REQUESTED** [review]

**Review by pashidlos — APPROVED** [review]
LGTM!

**Comment by pashidlos** [comment]
@suratdas thanks for contribution!

**Pull request merged**